feat: add `iter/cartesian-power`

TheNourhan opened the pull request:
Resolves #1336

## Description

> What is the purpose of this pull request?

The purpose of this pull request is to implement the iterCartesianPower function, which generates the Cartesian power of an input array-like object.

## This pull request:

- Implementation of the iterCartesianPower function in the `lib` folder.
- Documentation updates in the `docs` folder, including:
  - Writing REPL usage examples in the `docs/repl.txt` file.
  - Writing TypeScript type definitions in the `docs/types` folder.
- Writing a README.md file for the iterCartesianPower function.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #1336

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

TheNourhan committed:
add @stdlib/iter/cartesian-power

kgryte commented:
@TheNourhan Would you mind checking the box in the OP indicating that you've read the project contributing guidelines? Thanks!

kgryte reviewed (changes requested):
Thanks for working on this @TheNourhan. Looks like this feature addition is missing tests and benchmarks. We won't be able to review until those are added. Thanks!

kgryte commented:
Also, please ensure that you have run `make init` and have installed EditorConfig. Currently, your contribution includes a number of lint errors and uses spaces rather than tabs in JavaScript files.

TheNourhan commented:
Can you tell me the errors in any file, please?

kgryte commented on node_modules/@stdlib/iter/cartesian-power/lib/main.js:
```diff
@@ -0,0 +1,147 @@
+*/
+
+function iterCartesianPower(x, n) {
+    var iter;
```
This should be tab indentation. Make sure all JS files use tab indentation.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/lib/main.js:
```diff
@@ -0,0 +1,147 @@
+
+function iterCartesianPower(x, n) {
+    var iter;
+    var FLG = false;
+    var i = 0;
+    var j = [];
+    var idx = [];
```
This should have triggered lint errors. Please make sure you have run `make init` before attempting to commit code.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/lib/main.js:
```diff
@@ -0,0 +1,147 @@
+    var idx = [];
+
+    // Validate input arguments
+    if (!Array.isArray(x) && !isTypedArray(x) && !isIterable(x)) {
```
Please have a look through the codebase and match our spacing conventions. Additionally, you should not be checking for iterables, as you are not handling iterables in the implementation. Lastly, check for a collection, rather than using `isArray` and `isTypedArray`.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/lib/main.js:
```diff
@@ -0,0 +1,147 @@
+
+    // Validate input arguments
+    if (!Array.isArray(x) && !isTypedArray(x) && !isIterable(x)) {
+        throw new TypeError('invalid argument. First argument must be an array-like object.');
```
Include the value in the error message, similar to L65.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/lib/main.js:
```diff
@@ -0,0 +1,147 @@
+    if (!Array.isArray(x) && !isTypedArray(x) && !isIterable(x)) {
+        throw new TypeError('invalid argument. First argument must be an array-like object.');
+    }
+    if (!isNumber(n) || isnan(n)) {
```
This check is not necessary given L67.

kgryte commented:
This PR is still missing tests and benchmarks.

TheNourhan committed:
update cartesian-power

TheNourhan commented:
I used npm init, the Editorkonfig format and added Testes & Benchmarks, and updated the files you mentioned.
The branch that I pulled doesn't show me errors.

Thanks for your feedback.

TheNourhan committed:
Updating spaces

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
```diff
@@ -0,0 +1,212 @@
+
+# iterCartesianPower
+
+> Create an iterator which generates the Cartesian power of an input array-like object.
```
```suggestion
> Create an iterator which returns the Cartesian power.
```

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
```diff
@@ -0,0 +1,212 @@
+
+#### iterCartesianPower( x, n )
+
+Returns an iterator which generates the Cartesian power of an input array-like object.
```
```suggestion
Returns an iterator which returns the Cartesian power.
```

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
```diff
@@ -0,0 +1,212 @@
+var n = 2;
+var iterator = iterCartesianPower( x, n );
+
+for (const combination of iterator) {
```
Everything should be written in ES5. See also https://github.com/stdlib-js/stdlib/blob/develop/FAQ.md#es2015-and-beyond.

kgryte replied:
You can perform manual iteration. See other iterator packages for examples.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
````diff
@@ -0,0 +1,212 @@
+Returns an iterator which generates the Cartesian power of an input array-like object.
+
+```javascript
+var x = ['a', 'b', 'c'];
+var n = 2;
+var iterator = iterCartesianPower( x, n );
````
```suggestion
var x = [ 'a', 'b', 'c' ];

var iterator = iterCartesianPower( x, 2 );
```

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
```diff
@@ -0,0 +1,212 @@
+
+<section class="notes">
+
+## Notes
+
+- The function expects the first argument x to be an array-like object.
+- The second argument n must be a non-negative integer.
+- The returned iterator will generate all possible combinations of n elements from the input array x.
+    
```
```suggestion
```

We don't typically explain edge cases. For the last note, this is implied by the concept of the Cartesian power.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
```diff
@@ -0,0 +1,212 @@
+<section class="examples">
+
+## Examples
+
```
```suggestion
```

Do not include spurious newlines unless specified to do so by linting.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
````diff
@@ -0,0 +1,212 @@
+
+```
+
+```javascript
````
Don't split up examples like this. This does not match the conventions in this project.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
```diff
@@ -0,0 +1,212 @@
+var n = 2;
+var iterator = iterCartesianPower(x, n);
+
+for (const combination of iterator) {
```
ES5 here and elsewhere.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
```diff
@@ -0,0 +1,212 @@
+
+<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->
+
+<section class="related">
+
+* * *
+
+## See Also
+
+-   <span class="package-name">[`@stdlib/array/from-iterator`][@stdlib/array/from-iterator]</span><span class="delimiter">: </span><span class="description">create (or fill) an array from an iterator.</span>
+-   <span class="package-name">[`@stdlib/iter/datespace`][@stdlib/iter/datespace]</span><span class="delimiter">: </span><span class="description">create an iterator which returns evenly spaced dates over a specified interval.</span>
+-   <span class="package-name">[`@stdlib/iter/incrspace`][@stdlib/iter/incrspace]</span><span class="delimiter">: </span><span class="description">create an iterator which returns evenly spaced numbers according to a specified increment.</span>
+-   <span class="package-name">[`@stdlib/iter/logspace`][@stdlib/iter/logspace]</span><span class="delimiter">: </span><span class="description">create an iterator which returns evenly spaced numbers on a log scale.</span>
+-   <span class="package-name">[`@stdlib/iter/step`][@stdlib/iter/step]</span><span class="delimiter">: </span><span class="description">create an iterator which returns a sequence of numbers according to a specified increment.</span>
+-   <span class="package-name">[`@stdlib/iter/unitspace`][@stdlib/iter/unitspace]</span><span class="delimiter">: </span><span class="description">create an iterator which returns numbers incremented by one.</span>
```
```suggestion
<section class="related">
```

kgryte commented on node_modules/@stdlib/iter/cartesian-power/README.md:
```diff
@@ -0,0 +1,212 @@
+
+[@stdlib/math/base/special/roundn]: https://www.npmjs.com/package/@stdlib/math-base-special-roundn
+
+<!-- <related-links> -->
+
+[@stdlib/array/from-iterator]: https://www.npmjs.com/package/@stdlib/array-from-iterator
+
+[@stdlib/iter/datespace]: https://github.com/stdlib-js/iter/tree/main/datespace
+
+[@stdlib/iter/incrspace]: https://github.com/stdlib-js/iter/tree/main/incrspace
+
+[@stdlib/iter/logspace]: https://github.com/stdlib-js/iter/tree/main/logspace
+
+[@stdlib/iter/step]: https://github.com/stdlib-js/iter/tree/main/step
+
+[@stdlib/iter/unitspace]: https://github.com/stdlib-js/iter/tree/main/unitspace
+
+<!-- </related-links> -->
+
```
```suggestion
```

kgryte commented on node_modules/@stdlib/iter/cartesian-power/benchmark/benchmark.js:
```diff
@@ -0,0 +1,49 @@
+
+// MODULES //
+
+const bench = require( '@stdlib/bench' );
```
Please run linting over this file. See https://github.com/stdlib-js/stdlib/tree/develop/tools/make/lib/lint#lint-javascript-benchmarks.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/benchmark/benchmark.js:
```diff
@@ -0,0 +1,49 @@
+	// Run the benchmark...
+	b.tic();
+	for ( var i = 0; i < b.iterations; i++ ) {
+		v = iterator.next();
+		if ( v.done ) {
+			b.fail( 'should iterate over all elements' );
+		}
+	}
+	b.toc();
+	if ( v.done ) {
+		b.pass( 'benchmark finished' );
+	} else {
+		b.fail( 'should have finished iteration' );
```
This logic is not correct. You actually need to ensure that `v.done` does **not** return `true`. The point is to test per iteration performance, not how long it can generate all combinations. If folks want to generate all combos, they should just use `@stdlib/array/cartesian-power`. Instead, we want to know how long it takes a single iteration.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/docs/repl.txt:
```diff
@@ -0,0 +1,36 @@
+
+    Parameters
+    ----------
+     x : ArrayLike<T> | Iterable<T>
```
This is not TypeScript. Please read https://github.com/stdlib-js/stdlib/blob/develop/docs/repl_text.md

kgryte commented on node_modules/@stdlib/iter/cartesian-power/lib/main.js:
```diff
@@ -0,0 +1,153 @@
+			res.push( x[ idx[ k ] ] );
+		}
+		// Update indices
+		for ( var k = n - 1; k >= 0; k-- ) {
```
We explicitly hoist declared variables. See other packages for examples.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/package.json:
```diff
@@ -0,0 +1,70 @@
+    "utility",
+    "utils",
+    "util",
+    "linspace",
+    "linear",
+    "sequence",
+    "monotonic",
+    "arange",
+    "range",
```
```suggestion
    "cartesian",
    "combination",
    "permutation",
```

kgryte commented on node_modules/@stdlib/iter/cartesian-power/test/test.js:
```diff
@@ -0,0 +1,56 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+'use strict';
```
Please match the spacing and style conventions of other packages in the project.

kgryte commented on node_modules/@stdlib/iter/cartesian-power/test/test.js:
```diff
@@ -0,0 +1,56 @@
+tape( 'the function returns an iterator which generates all possible combinations of length `n` containing elements from the input array', function test( t ) {
+	var expected = [ [ 'a', 'a' ], [ 'a', 'b' ], [ 'a', 'c' ], [ 'b', 'a' ], [ 'b', 'b' ], [ 'b', 'c' ], [ 'c', 'a' ], [ 'c', 'b' ], [ 'c', 'c' ] ];
+	var actual = [];
+	for ( var v of iterCartesianPower( [ 'a', 'b', 'c' ], 2) ) {
```
ES5.

kgryte commented:
@TheNourhan In general, I suggest trying to make your contribution look as closely to other packages as possible. We author everything in ES5 and we enforce various lint conventions, which are not currently adhered to.

TheNourhan committed:
Updating README and writing in ES5

TheNourhan commented:
I made all the changes you mentioned except for the benchmark file and lint errors I will work on them after I fix the installation errors in my project because the make command does not work well in my env.
Thanks for your feedback.

TheNourhan committed:
Updating index.d.ts example

TheNourhan committed:
Updating benchmark.js

TheNourhan committed:
Updating Copyright

TheNourhan commented:
I've made all the changes you mentioned

kgryte commented:
@TheNourhan Great! Thanks for the update. We'll try and review within the next day or so.

TheNourhan commented:
I'm going to close this PR and open a new one, I reinstalled the project and fixed all the errors

kgryte commented:
@TheNourhan Sounds good. Feel free to do so.

kgryte commented:
@TheNourhan Is this PR obsolete? Based on your prior comments, you were planning on submitting a new PR.

TheNourhan commented:
@kgryte  I've done it here https://github.com/stdlib-js/stdlib/pull/2534
